packages/synapse-react-client/src/components/DataGrid/DataGrid.tsx

```diff
@@ -0,0 +1,695 @@
+  const synapseClient = useSynapseContext().synapseClient
+
+  // Grid session state
+  const [sessionId, setSessionId] = useState<string>('')
+  const [replicaId, setReplicaId] = useState<number | null>(null)
+  const [presignedUrl, setPresignedUrl] = useState<string>('')
+  const [gridSql, setGridSql] = useState<string>('')
+  const inputRef = useRef<HTMLInputElement>(null)
+  const webSocketRef = useRef<WebSocket | null>(null)
+
+  // WebSocket state
+  const [isConnected, setIsConnected] = useState<boolean>(false)
+  const [connectionStatus, setConnectionStatus] =
+    useState<string>('Disconnected')
+  const [messages, setMessages] = useState<string[]>([])
+
+  // CRDT state
+  const sequenceNumberRef = useRef<number>(1)
+  const modelRef = useRef<Model | null>(null)
+  interface ModelSnapshot {
+    columnNames: string[]
+    columnOrder: number[]
+    rows: string[]
+  }
+  interface jsonJoyMessage {
+    sequenceNumber: number
+    methodName: string
+    payload?: string
+  }
+  const modelSnapshotRef = useRef<ModelSnapshot>({
+    columnNames: [],
+    columnOrder: [],
+    rows: [],
+  })
+  const getModel = () => modelRef.current
+  const setModel = (newModel: Model) => {
+    if (newModel.view() !== modelRef.current?.view()) {
+      modelRef.current = newModel
+      modelSnapshotRef.current = modelRef.current.api.getSnapshot()
+    }
+  }
+
+  // If grid sessionId or replicaId changes, reset the model
+  useEffect(() => {
+    if (sessionId || replicaId) {
+      modelRef.current = null
+      modelSnapshotRef.current = {
+        columnNames: [],
+        columnOrder: [],
+        rows: [],
+      }
+      setRowValues([])
+      setColValues([])
+      setGridReady(false)
+      sequenceNumberRef.current = 1
+      setMessages([]) // Clear message log
+      setIsConnected(false)
+      setConnectionStatus('Disconnected')
+      if (webSocketRef.current) {
+        webSocketRef.current.close()
+        webSocketRef.current = null
+      }
+    }
+  }, [sessionId, replicaId])
+
+  // Grid rows and columns
+  type DataGridRow = { [key: string]: string | number }
+  const [gridReady, setGridReady] = useState<boolean>(false)
+  const [rowValues, setRowValues] = useState<DataGridRow[]>(
+    modelRowsToGrid(modelSnapshotRef.current),
+  )
+  const [colValues, setColValues] = useState<Column[]>(
+    modelColsToGrid(modelSnapshotRef.current),
+  )
+  const [prevRows, setPrevRows] = useState(rowValues)
+
+  // Query Input can either be an empty string, a SQL query, or a session ID
+  const parseQueryInput = (input: string) => {
+    const trimmedInput = input.trim()
+    if (!trimmedInput) {
+      return { type: 'empty', input: '' }
+    } else if (trimmedInput.toUpperCase().startsWith('SELECT')) {
+      return { type: 'sql', input: trimmedInput }
+    } else if (/^N\w*=$/.test(trimmedInput)) {
+      return { type: 'sessionId', input: trimmedInput }
+    }
+    return { type: 'unknown', input: trimmedInput }
+  }
```

Comment:
Static utilities could be moved into a new file `DataGridUtils.ts` which would simplify this component, and those utilities could be tested in isolation.